### PR TITLE
Add support for multiple inputs to out_wrapper and strict dtype checking

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #78616
* #78615
* #78350
* #80219
* #80443
* __->__ #79941
* #80516

When a function returns multiple parameters in PyTorch, the `out`
parameter takes a tuple of tensors (see `linalg.svd` for example).
The current implementation in `out_wrapper_multi` modelled this wrong,
as it assumed that it would take a number of different named
parameters.

This PR implements the correct behaviour in `out_wrapper`. As a small
side-effect, we now need to call `@out_wrapper()` when the output is
just one tensor.

This PR also implements an additional optional parameter that checks
whether the dtype of the given `out` is exactly the dtype that the meta
function requires. This is the behaviour that we currently have in
PyTorch, and this check is necessary in eager when we call with these
tensors into external libraries.

We also make the functions with several outputs return a namedtuple,
similar to what we do in PyTorch.

### CODE REVIEW

**Review by mruberry — APPROVED** [review]
Ship it!

**Comment by Lezcano** [comment]
@ezyang This PR is failing on a meta test, but I'm failing to track down where is the issue coming from. Now, I think it's coming from some torch_function modification you did. Any guesses?

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -323,9 +323,8 @@ def meta_index_Tensor(self, indices):
     return self.new_empty(before_shape + replacement_shape + after_shape)
 
 
-@torch.library.impl(meta_lib, "addbmm")
```
Comment (ezyang) [inline torch/_meta_registrations.py:326]: Doesn't this explain why the test is failing?

Comment (Lezcano) [reply torch/_meta_registrations.py:326]: That was indeed wrong (corrected) but was not the cause of the error. The thing is that somewhere the `out` parameter (which is a tuple) is being transformed into several kwargs, each with the name of the corresponding output. This is wrong, and I have not been able to track down where it happens, but it seems related to the contextmanager `enable_torch_function_mode(inner)`...

To see the backtrace of the failures see https://github.com/pytorch/pytorch/runs/7062769044?check_suite_focus=true

Comment (ezyang) [reply torch/_meta_registrations.py:326]: The function in question here is being registered as an ATen op directly in the dispatcher. Theseget invoked by the function: parseIValuesToPyArgsKwargs

```
  // Populate positional arguments
  for (const auto idx : c10::irange(positional_default_start)) {
    PyTuple_SET_ITEM(
        args.ptr(), idx, schemaAwareToPyObject(idx).release().ptr());
  }
      
  // Populate keyword arguments
  for (const auto idx : c10::irange(kwarg_only_start, arguments.size())) {
    // But don't populate default keyword arguments
    if (is_default(idx))
      continue;
    const auto& arg = schema.arguments()[idx];
    kwargs[py::cast(arg.name())] = schemaAwareToPyObject(idx);
  }   
  return std::make_pair(std::move(args), std::move(kwargs));
```

There is no special handling for out arguments. The dispatcher just slaps them exactly as they look in native_functions.yaml

I'm not convinced I want to change this behavior. But that means we will need two different implementations of out handling, one for refs (which are Python visible) and another for the ATen registrations (which will see them unpacked.)

Comment (Lezcano) [reply torch/_meta_registrations.py:326]: Couldn't this be solved just by packing the `out` parameters back into a tuple in that context manager before feeding them back to the ref?

Comment (ezyang) [reply torch/_meta_registrations.py:326]: Yes, but the main point is you need two versions of the function, one to be a ref and one to actually register to the dispatcher

File: torch/_decomp/__init__.py [hunk 2]
```diff
@@ -37,7 +40,49 @@ def clamp_min(x):
     `disable_meta` to disable this behavior.
     """
 
-    def decomposition_decorator(f):
+    def decomposition_decorator(f: Callable) -> Callable:
+        sig = inspect.signature(f)
```
Comment (Lezcano) [inline torch/_decomp/__init__.py:44]: @rgommers all this block of code is the juggling with `inspect` and annotations I told you about. given a function that takes a kwargs `out=` which is a tuple of tensors and returns a namedtuple, it wraps it as a function that takes the each of the arguments as a separate tensor.

The dirty parts are:
- Accessing `out_annotations.__origin__` to check that that type is indeed a `Tuple` 
- Accessing `return_annotation._fields` to get the names from the `NamedTuple` from the return.

**Review by ezyang — APPROVED** [review]

File: torch/_decomp/__init__.py (hunk 2 above)
Comment (rgommers) [reply torch/_decomp/__init__.py:44]: Thanks @Lezcano, interesting. This is more like codegen (new annotation from old annotation) than actual use of type annotations at runtime to do anything functional, so my concern of using annotations at runtime (which is a no no) doesn't really apply. I'm not surprised that there's no clean ways in the language or in `typing`/`typing_extensions` to do this kind of thing.

**Pull request closed without merging**